File: Sorting Number/MergeSort.java
public class MergeSort {
    public static void mergeSort(int[] num, int n){
        if(n < 2){
            return;
        }
        int mid = n / 2;
        int[] l = new int[mid];
        int[] r = new int[n - mid];
        for(int i = 0; i < mid; i++){
            l[i] = num[i];
        }
        for(int i = mid; i < n; i++){
            r[i - mid] = num[i];
        }
        mergeSort(l, mid);
        mergeSort(r, n - mid);

        merge(num, l, r, mid, n - mid);
    }
    public static void merge(int[] num, int[] l, int[] r, int left, int right){
        int i = 0, j = 0, k = 0;
        while(i < left && j < right){
            if(l[i] <= r[j]){
                num[k++] = l[i++];
            }else{
                num[k++] = r[j++];
            }
        }
        while(i < left){
            num[k++] = l[i++];
        }
        while(j < right){
            num[k++] = r[j++];
        }
    }
}

File: Sorting Number/Merge Sort.py
# done recursively, using divide and conquer strategy
# the function does in an inplace sort, hence there is no need for a return value
def mergesort(nlist):
    if len(nlist) > 1: # checks if the length of list is greater than 1
        mid = len(nlist)//2 # identify the middle element for the split
        llist = nlist[:mid]
        rlist = nlist[mid:]

        mergesort(llist) # call merger sort recursively on the split lists
        mergesort(rlist)
        i=j=k=0
        while i < len(llist) and j < len(rlist): # check if the counter has not exceeded the length of the lists
            if llist[i] < rlist[j]: # compare the values between the 2 lists
                nlist[k] = llist[i] # append or add smaller value to new list
                i+=1
            else:
                nlist[k] = rlist[j]
                j+=1
            k+=1
        while i < len(llist): # add all remaining elements in left list to resultant list
            nlist[k] = llist[i] 
            k+=1
            i+=1
        while j < len(rlist): # add all remaining elements in right list to resultatnt list
            nlist[k] = rlist[j]
            k+=1
            j+=1
    else:
        nlist = nlist # if lenght of list is less than 1
